CIP-XXXX/CIP-XXXX.md

```diff
@@ -0,0 +1,180 @@
+
+This specification defines the manner which javascript code is injected into webpages, accessed by the webpage/dApp, as well as defining the API for dApps to communicate with the user's wallet.
+
+## Motivation: why is this CIP necessary?
+
+This *base* API intends to standardize the initial dApp-Wallet connection while defining a path for how functionality can be enhanced through future CIP extensions. All wallet providers are expected to support this API with future extensions supported at the wallet's discretion.
+
+[CIP-30 | Cardano dApp-Wallet Web Bridge](https://github.com/cardano-foundation/CIPs/tree/master/CIP-0030) facilitated the launch of dApp development on Cardano. Yet it's functionality is limited in scope leading to multiple revisions of the specification. Causing divergence between capabilities of wallet implementations, making the development of compatible dApps cumbersome and error-prone.
+
+This proposal aims to act as a skeleton from which functionality can be fleshed out in a modular fashion. With the first supported extension containing CIP-30's `Full API` functionality. This proposal lifts the `Initial API` from CIP-30 and places it here. This is an attempt to improve the backwards compatibility of this proposal as by implementing this proposal no CIP-30 functionality is lost.
+
+This proposal is intended for dApp creators and Cardano wallet providers, with broad application to various types of Web3 applications.
+
+## Specification
+
+### Data Types
+
+#### Extension
+
+An extension is an object with a single field `"cip"` that describe a CIP number extending the API (as a plain integer, without padding). For example:
+
+```
+{ "cip": 30 }
+```
+##### Known Extensions
+
+This should act as the definitive list of known and accepted extensions, all extensions which have  been processed through the CIP process should be added here. 
+
+| CIP    | Scope                           |
+| ------ | ------------------------------- |
+| CIP-30 | Base interface                  |
+
+##### Can extensions add their own datatypes?
+
+Yes.
+
+##### Can extensions extend the initial API?
+
+No. Only if this specification itself if replaced.
+
+##### Can extensions depend on other extensions?
+
+Yes. Extensions may have other extensions as pre-requisite. Some newer extensions may also invalidate functionality introduced by earlier extensions. There's no particular rule or constraints in that regards. Extensions are specified as CIP, and will define what it entails to enable them.
+
+##### Should extensions follow a specific format?
+
+Yes. They all are CIPs.
+
+##### Can extensions add their own endpoints and/or error codes?
+
+Yes. Extensions may introduce new endpoints or error codes, and modify existing ones. Extensions may even change the rules outlined in this very proposal. The idea being that wallet providers should start off implementing this CIP, and then walk their way to implementing their chosen extensions.

```

Comment:
> This could be a mess.

I agree, we need to navigate this carefully. 

I see three options:
- we silo extensions, not allowing them to overlap or share anything (datatypes, error codes, etc.) only allowing them to extend this API. I believe this aligns with your suggestion.
-  as described currently - we allow extensions to share, replace and overlap with each other.
- somewhere in the middle, I think some novel design would be needed for this to be viable.

I would like to hear other's insights on how is best to handle this.

Reply:
I lean toward option 1 on your list. It could quickly become a mess with dependencies between extensions etc. If the same data types/error codes are defined in multiple extensions, it could be up to the implementer to combine and merge as it sees fit.

Reply:
> I lean toward option 1 on your list.

I have been reflecting on this and I agree. We can always give extensions more freedom to share down the line, if needed. What we cannot do is try to impose restrictions on extensions once we discovery that granting flexibility has resulted in a dependency nightmare.

Reply:
This change has been made.